src/OrleansRuntime/Catalog/Catalog.cs

```diff
@@ -632,46 +632,43 @@ private void CreateGrainInstance(string grainTypeName, ActivationData data, stri
 
             GrainTypeData grainTypeData = GrainTypeManager[grainClassName];
 
+            //Get the grain's type
             Type grainType = grainTypeData.Type;
 
-            //Create a new instance of the given grain type
-            var grain = grainCreator.CreateGrainInstance(grainType, data.Identity);
-            grain.Data = data;
-
+            //Gets the type for the grain's state
             Type stateObjectType = grainTypeData.StateObjectType;
-            object state = null;
-            if (stateObjectType != null)
-            {
-                state = Activator.CreateInstance(stateObjectType);
-            }
             
             lock (data)
             {
-                data.SetGrainInstance(grain);
-                var statefulGrain = grain as IStatefulGrain;
-                if (statefulGrain != null)
+                Grain grain;
+
+                //Create a new instance of a stateless grain
+                if (stateObjectType == null)
                 {
-                    if (state != null)
-                    {
-                        SetupStorageProvider(data);
-                        statefulGrain.GrainState.State = state;
-                        statefulGrain.SetStorage(new GrainStateStorageBridge(data.GrainTypeName, statefulGrain,
-                            data.StorageProvider));
-                    }
+                    //Create a new instance of the given grain type
+                    grain = grainCreator.CreateGrainInstance(grainType, data.Identity);
```

Comment:
@jdom The reason for this was to ensure that data.Identity & data.StorageProvider are within a lock as it was in the original. However, looking at data.Identity, this should be safe no in the lock since it appears to be readonly that is set in the constructor for `ActivationAddress`. data.StorageProvider is only set in `private void SetupStorageProvider(Type grainType, ActivationData data)` within `Catalog' and there is no lock there at all.

Please let me know what you think is best here.